src/Chapter3.hs

```diff
@@ -1008,6 +1108,51 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+newtype Damage = Damage Int
+newtype Defence = Defence Int
+newtype Life = Life Int
+
+data HolyKnight = HolyKnight { kDamage:: Damage, kDefence:: Defence, kLife:: Life }
+data DarkMonster = DarkMonster { mDamage:: Damage, mLife:: Life }
+
+data KnightAction = KnightAttack | CastSpell Int | DrinkHealth Int
+data MonsterAction = MonsterAttack | Run
+
+class AttackMove a where
+  damageTaken:: a -> Damage -> a
+  attackDone:: a -> Damage
+
+instance AttackMove HolyKnight where
+  damageTaken k@(HolyKnight _ (Defence d) (Life h)) (Damage x) = k { kLife = Life (h - max (x - d) 0) }
+  attackDone (HolyKnight x _ _) = x
+
+instance AttackMove DarkMonster where
+  damageTaken d@(DarkMonster _ (Life h)) (Damage x) = d { mLife = Life (h - x) }
+  attackDone (DarkMonster x _) = x
+
+data Turn = KnightsTurn | MonstersTurn
+
+battle:: HolyKnight -> [KnightAction] -> DarkMonster -> [MonsterAction] -> Bool
+battle holyKnight knightActions darkMonster darkMonsterActions =
```

Comment:
That's a correct implementation! 👍🏻 
However, your function only works for fighting a knight with a monster. But not for two knights or two monsters. To support these cases, you will need to implement more general abstractions. But your solution is already quite good 🙂